File: leetcode/p0001/solution.py
# -*- coding: utf-8 -*-


class Solution:

    def twoSum(self, nums: list, target: int) -> list:
        cache = {}
        for i, num in enumerate(nums):
            if num in cache and i != cache[num]:
                return [i, cache[num]]
            cache[target - num] = i
        return []

    def twoSum2(self, nums: list, target: int) -> list:
        objs = [(i, num) for i, num in enumerate(nums)]
        objs.sort(key=lambda x: x[1])
        low, high = 0, len(nums) - 1
        while low < high:
            diff = target - objs[low][1]
            if diff > objs[high][1]:
                while low + 1 < high and objs[low][1] == objs[low + 1][1]: low += 1
                low += 1
            elif diff < objs[high][1]:
                while high - 1 >= low and objs[high][1] == objs[high - 1][1]: high -= 1
                high -= 1
            else:
                return [objs[low][0], objs[high][0]]
        return []


if __name__ == '__main__':
    inputs = [
        ([2, 7, 11, 15], 9)
    ]

    for nums, target in inputs:
        print(' Given nums: ', nums, ', target = ', target, sep='')
        solution = Solution()
        pos = solution.twoSum(nums, target)
        print(' Output 1: ', pos, sep='')
        pos = solution.twoSum2(nums, target)
        print(' Output 2: ', pos, sep='')

File: leetcode/p0001/Solution.java
import java.util.Arrays;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import static java.util.Comparator.*;

public class Solution {
    public int[] twoSum(int[] nums, int target) {
        List<Pair> objs = new ArrayList<>();
        for (int i = 0; i < nums.length; i++) {
            objs.add(new Pair(i, nums[i]));
        }
        //System.out.println(Arrays.toString(objs.toArray()));
        Collections.sort(objs);
        //System.out.println(Arrays.toString(objs.toArray()));
        int low = 0, high = objs.size() - 1;
        while (low < high) {
            int diff = target - objs.get(low).value;
            if (diff > objs.get(high).value) {
                while ((low + 1 < high)
                        && (objs.get(low).value == objs.get(low + 1).value)) {
                    low++;
                }
                low++;
            } else if (diff < objs.get(high).value) {
                while ((high - 1 >= low)
                        && (objs.get(high).value == objs.get(high - 1).value)) {
                    high--;
                }
                high--;
            } else {
                return new int[] {objs.get(low).index, objs.get(high).index};
            }
        }
        return null;
    }

    static final class Pair implements Comparable<Pair> {
        // Comparable wih comparator construction methods
        private static final Comparator<Pair> COMPARATOR =
            comparingInt((Pair p) -> p.value);

        private final int index;
        private final int value;

        Pair(int index, int value) {
            this.index = index;
            this.value = value;
        }

        @Override
        public int compareTo(Pair p) {
            return COMPARATOR.compare(this, p);
        }

        @Override
        public String toString() {
            return "(index=" + index + ", value=" + value + ")";
        }
    }

    public int[] twoSum2(int[] nums, int target) {
        Map<Integer, Integer> map = new HashMap<>();

        for (int i = 0; i < nums.length; i++) {
            if (map.containsKey(nums[i]) && i != map.get(nums[i])) {
                return new int[] {i, map.get(nums[i])};
            }
            map.put(target - nums[i], i);
        }

        return null;
    }

    public static void main(String[] args) {
        Input[] inputs = {
            new Input(new int[]{ 2, 7, 11, 15 }, 9),
        };

        for (int i = 0; i < inputs.length; i++) {
            int[] nums = inputs[i].nums;
            int target = inputs[i].target;

            System.out.printf("\n Given nums: %s, target = %d",
                    Arrays.toString(nums), target);

            Solution solution = new Solution();
            int[] pairs = solution.twoSum(nums, target);
            System.out.printf("\n Output: %s\n",
                    (pairs == null) ? "Not found" : Arrays.toString(pairs));
        }
    }

    static final class Input {
        private final int[] nums;
        private final int target;

        Input(int[] nums, int target) {
            this.nums = nums;
            this.target = target;
        }
    }
}
